search/arch_draft.py

```diff
@@ -0,0 +1,82 @@
+"""
+This is temporary module with the arch concept for the search.
+
+@todo #331:60m  Implement the new arch at the search app.
+ All you need is to implement ContextsStack class
+ and place the other classes to the relevant places.
+ `search.search` module will be fully rewritten.
+ You also can drop away starting_by_term functionality.
+ See "is_name_start_by_term" aggregation parameter
+ inside search.search.search function.
+"""
+
+import abc
+from itertools import chain
+
+from django.db.models import QuerySet
+from django.http import JsonResponse
+
+from images.models import Image
+from pages import typing
+
+
+class Representable(abc.ABC):
+    @property
+    def name(self):
+        raise NotImplemented()
+
+    @property
+    def url(self):
+        raise NotImplemented()
+
+
+class Result:
+    def __init__(self, item: Representable):
+        self.name: str = item.name
+        self.url: str = item.url
+        self.image: Image = getattr(item, 'image', None)
+        self.price: float = getattr(item, 'price', 0.0)
+
+
+class Results:
+    def __init__(self, results_qs: QuerySet):
+        """`results_qs` is QuerySet with already searched and sorted entities."""
```

Comment:
>searched

We have complex [search strategies](https://github.com/fidals/refarm-site/blob/master/search/search.py#L103). I don't find related logic here, where will we move it?

Reply:
@artemiy312 , we'll use [django orm search](https://docs.djangoproject.com/en/1.11/topics/db/search/) mech.
I wrote about it at the comment at autocomplete view below:
```
def autocomplete(request, query):
    # we'll use django orm search based on postgresql.
    # Query sets will already contain searched entities.
```

Reply:
We already have used it, but it is still a complex search strategy, that should be generalized somehow as it is in current implementation

Reply:
@artemiy312 , current proposal covers preparing search results for a front side with a clear contract.
It does not cover search itself. It just assumes that all search logic will be implemented with QuerySet objects since it's django way.

Why it does not cover search engine arch? Because it's fully designed at the Django orm side. Just implementation is left here